tests/unit/test_cell_operations_service.py

```diff
@@ -2,28 +2,301 @@
+        ).thenRaise(dm_exc.NotFound())
+        with pytest.raises(exceptions.InvalidInputError):
+            self.cell_operations_service.authorize_model_for_user(user=USER_NON_ADMIN, model_name="non_existent_model")
+
+    def test_mismatched_embeddings_and_queries(self) -> None:
+        embeddings = [[0, 1, 2]]
+        knn_response = MatchResult(matches=[])
+        self.__mock_apis(
+            model=MODEL,
+            index=INDEX,
+            anndata_data=ANNDATA_DATA,
+            embeddings=embeddings,
+            matching_client_response=knn_response,
+        )
+        with pytest.raises(
+            exceptions.VectorSearchResponseError,
+            match=re.escape("Number of query ids (1) and knn matches (0) does not match."),
+        ):
+            CellOperationsService._CellOperationsService__validate_knn_response(
+                embeddings=embeddings, knn_response=knn_response
+            )
+
+    def test_empty_neighbors(self) -> None:
+        embeddings = [[0, 1, 2]]
+        knn_response = MatchResult(matches=[MatchResult.NearestNeighbors(neighbors=[])])
+        self.__mock_apis(
+            model=MODEL,
+            index=INDEX,
+            anndata_data=ANNDATA_DATA,
+            embeddings=embeddings,
+            matching_client_response=knn_response,
+        )
+        with pytest.raises(
+            exceptions.VectorSearchResponseError, match=re.escape("Vector Search returned a match with 0 neighbors.")
+        ):
+            CellOperationsService._CellOperationsService__validate_knn_response(
+                embeddings=embeddings, knn_response=knn_response
+            )
+
+    @parameterized.expand(
+        [
+            ([], [], False),
+            ([[0, 1, 2]], ["erythrocyte"], False),
+            ([[0, 1, 2], [3, 4, 5]], ["erythrocyte", "monocyte"], True),
+        ]
+    )
+    @pytest.mark.asyncio
+    async def test_annotate_adata_file(
+        self, embeddings: List[List[float]], cell_types: List[str], include_dev_metadata: bool
+    ) -> None:
+        """
+        Test the annotate_adata_file method.
+
+        :param embeddings: The embeddings to use for the test.
+        :param cell_types: The cell types to use for the test.  Should match 1:1 with the embeddings array.
+        :param include_dev_metadata: Whether or not to set the include_dev_metadata flag when calling the method.
+        """
+        matching_client_response = self.__mock_apis(
+            model=MODEL, index=INDEX, anndata_data=ANNDATA_DATA, embeddings=embeddings
+        )
+
+        # mock calls to get cell distribution
+        temp_table_fqn = "temp_table_fqn"
+        query_ids = [f"q{i}" for i in range(len(embeddings))]
+        response = [
+            {"query_cell_id": query_ids[i], "matches": [{"cell_type": cell_types[i], "cell_count": 10}]}
+            for i in range(len(embeddings))
+        ]
+        when(self.cell_operations_service.cell_operations_dm).insert_matches_to_temp_table(
```

Comment:
I admittedly haven't used mockito in a while, so I might be misremembering and worrying over nothing.  What happens here if `insert_matches_to_temp_table` is called with any other inputs than the ones you've mocked up a response for?  Does it attempt to call the actual method instead of the mock, or does it still call the mock and then throw an error or something?

Reply:
Ah it does not because I mocked the two data manager classed (on line 45).  That's by I made the `cell_operations_dm` and `cellarium_general_dm` constructor parameters with defaults.  That lets the unit test pass in safe mocks that will return `None` if not stubbed out (just confirmed the behavior locally to be sure)